compiler/rustc_middle/src/mir/syntax.rs

```diff
@@ -512,6 +512,20 @@ pub struct CopyNonOverlapping<'tcx> {
     pub count: Operand<'tcx>,
 }
 
+/// Represents how a `TerminatorKind::Call` was constructed, used for diagnostics
+#[derive(Clone, Copy, TyEncodable, TyDecodable, Debug, PartialEq, Hash, HashStable)]
+#[derive(TypeFoldable, TypeVisitable)]
+pub enum CallDesugaring {
```

Comment:
Should this use a different name? Isn't desugaring exclusively used to refer to lowering of source-level constructs to internal constructs? So maybe `ThirCallSource` or something would be more appropriate?

Reply:
I'm not sure if `ThirCallSource` would be better as that means `Option<ThirCallSource>::None` means that there was no call source from THIR. `Option<CallDesugaring>::None` suggests that it was a normal call. For `CallDesugaring::Misc` they still didn't come from normal calls so it would still mean using calls internally to represent code that does not explicitly try to call functions

Reply:
Why can't we do this? 😅 

```
enum ThirCallSource {
  HirCall,
  OverloadedOperator,
  MatchCmp,
  Misc,
}
```